### ISSUE TITLE
Implied shortcut reference link produces spurious warning on nightly

### ISSUE THREAD

**Issue by ashtneoi** [rust-lang/rust#51231]
On nightly, when I run `cargo doc` or `rustdoc` on this crate...

`src/main.rs`:
```rust
pub mod bar {
    //! Here is the [`foo`] function.

    pub fn foo() { }
}

fn main() {
}
```

...it prints the following warning...

```none
warning: [foo] cannot be resolved, ignoring it...
```

...**even though the link is resolved and rendered correctly**. It happens regardless of whether `foo` is a function, struct, enum, trait, or module. I wasn't able to reproduce it on stable or beta, and it only seems to happen for `//!`, not `///`.

Interestingly, if I remove the `bar` module and put its contents in the crate root, or if I wrap `bar` in another module, the warning no longer appears (the link still works correctly in both cases).

## Meta

`rustc --version --verbose`:

```none
rustc 1.28.0-nightly (524ad9b9e 2018-05-29)
binary: rustc
commit-hash: 524ad9b9e03656f3fdeb03ed82fe78db3916e566
commit-date: 2018-05-29
host: x86_64-unknown-linux-gnu
release: 1.28.0-nightly
LLVM version: 6.0
```

**Comment by jyn514** [comment]
This works fine on current nightly, so I'm going to assume it's been fixed since last year. Feel free to re-open if you're still having trouble!
